OpenRA.Mods.Common/Traits/BotModules/HarvesterBotModule.cs

```diff
@@ -148,17 +167,93 @@ void IBotTick.BotTick(IBot bot)
+				return false;
+
+			// Tell the idle harvester to quit slacking:
+			var newSafeResourcePatch = FindNextResource(h.Actor, h);
+			AIUtils.BotDebug($"AI: Harvester {h.Actor} is idle. Ordering to {newSafeResourcePatch} in search for new resources.");
+			if (newSafeResourcePatch != Target.Invalid)
+				bot.QueueOrder(new Order("Harvest", h.Actor, newSafeResourcePatch, false));
+			else
+				h.NoResourcesCooldown = Info.ScanIntervalMultiplerWhenNoResources;
+
+			return true;
+		}
+
 		Target FindNextResource(Actor actor, HarvesterTraitWrapper harv)
 		{
-			bool IsValidResource(CPos cell) =>
-				harv.Harvester.CanHarvestCell(cell) &&
-				claimLayer.CanClaimCell(actor, cell);
+			// Prefer resource nearby to the nearest drop off point, otherwise scan from the current location.
+			var scanFromActor = harv.DockClientManager.ClosestDock(null, ignoreOccupancy: true)?.Actor ?? actor;
 
-			var path = harv.Mobile.PathFinder.FindPathToTargetCellByPredicate(
-				actor, new[] { actor.Location }, IsValidResource, BlockedByActor.Stationary,
-				loc => world.FindActorsInCircle(world.Map.CenterOfCell(loc), Info.HarvesterEnemyAvoidanceRadius)
-					.Where(u => !u.IsDead && actor.Owner.RelationshipWith(u.Owner) == PlayerRelationship.Enemy)
-					.Sum(u => Math.Max(WDist.Zero.Length, Info.HarvesterEnemyAvoidanceRadius.Length - (world.Map.CenterOfCell(loc) - u.CenterPosition).Length)));
+			var targets = resourceTypesByCell
+				.Where(kvp =>
+					harv.Harvester.Info.Resources.Contains(kvp.Value) &&
+					claimLayer.CanClaimCell(actor, kvp.Key))
+				.Select(kvp => kvp.Key);
+
+			var avoidanceCostForBin = new Dictionary<int2, int>();
+			var cellRadius = Info.HarvesterEnemyAvoidanceRadius.Length / 1024;
+			var minCellCost = harv.Mobile.Locomotor.Info.TerrainSpeeds.Values.Min(ti => ti.Cost);
+
+			static int2 CellToBin(CPos cell, int cellRadius)
+			{
+				return new int2(
+					cell.X / cellRadius,
+					cell.Y / cellRadius);
+			}
+
+			static int CalculateAvoidanceCostForBin(World world, int2 bin, int cellRadius, Actor actor, int minCellCost)
+			{
+				// Bins are overlapping, this allows actors to apply threat in both directions when they're at the edge.
+				// If the bins didn't overlap, actors along the edge of a bin only affect that bin, and not the bin next to it,
+				// despite the fact the are an equal risk to both.
+				var r = WDist.FromCells(cellRadius);
+				var vec = new WVec(r, r, WDist.Zero);
+				var originCell = new CPos(bin.X * cellRadius + cellRadius / 2, bin.Y * cellRadius + cellRadius / 2);
+				var origin = world.Map.CenterOfCell(originCell);
+				var threatActors = world.ActorMap.ActorsInBox(origin - vec, origin + vec)
+					.Where(u => !u.IsDead && actor.Owner.RelationshipWith(u.Owner) == PlayerRelationship.Enemy);
+
+				// For each actor in the threat radius, every cell we want to move is an extra 20x cost than a threat-free area.
+				return threatActors.Count() * minCellCost * 20;
```

Comment:
I suggest putting the `20` into the trait info. We already expose `PathingCost` when parsing `TerrainSpeeds` so I think it shouldn't alienate people. Maybe someone wants a harvester that is braver.